android/envoy/src/main/kotlin/org/greatfire/envoy/EnvoyState.kt

```diff
@@ -0,0 +1,118 @@
+    var concurrency = 6 // XXX
+
+    // moving this back to state because it's state
+    var connected = AtomicBoolean(false)
+    var activeServiceType = AtomicInteger(EnvoyServiceType.UNKNOWN.ordinal)
+    var activeService: EnvoyTest? = null
+    val additionalWorkingConnections = mutableListOf<EnvoyTest>()
+
+
+    // if set, wait an increasing amount of time before retrying blocked urls
+    var backoffEnabled = false
+
+    // our Cronet engine
+    var cronetEngine: CronetEngine? = null
+
+    // interface with application
+    var callback: EnvoyTestCallback? = null
+
+    // can we get away without this?
+    // maybe a callback method to return context?
+    var ctx: Context? = null
+
+    // Go library
+    val emissary = Emissary.newEmissary()
+
+    // DNS related code
+    val dns = EnvoyDns()
+
+    val shadowsocks: ShadowsocksService? = null
+
+    private fun createCronetEngine() {
+        // I think we can reuse the cache dir between runs?
+        // XXX we used to have multiple tests cronet based tests
+        // running in parallel...
+        val cacheDir = File(ctx!!.cacheDir, "cronet-cache")
+        if (!cacheDir.exists()) {
+            cacheDir.mkdirs()
+        }
+
+        cronetEngine = CronetNetworking.buildEngine(
+            context = ctx!!,
+            cacheFolder = cacheDir.absolutePath,
+            envoyUrl = null,
+            strategy = 0,
+            cacheSize = 10, // cache size in MB
+        )
+    }
+
+    // called when the connection worker found a successful connection
+    fun connectIfNeeded(test: EnvoyTest) {
+        // Check if we already have a working connection before continuing
+        // set that we do otherwise
+        if (connected.compareAndSet(false, true)) {
+
+            // start the cronet engine if we're using a cronet service
+            when (test.testType) {
+                EnvoyServiceType.CRONET_ENVOY,
+                EnvoyServiceType.CRONET_PROXY -> {
+                    createCronetEngine()
+                }
+                else -> "" // nothing to do
+            }
+
+            activeServiceType.set(test.testType.ordinal)
+            activeService = test
+            test.selectedService = true
+            Log.i(TAG, "🚀🚀🚀 Envoy connected")
+        } else if(test.testType == EnvoyServiceType.DIRECT) {
+            Log.i(TAG, "DIRECT overriding previous connection")
+
+            val previousService = activeService
+
+            // activate the direct connection
+            activeServiceType.set(test.testType.ordinal)
+            activeService = test
+            test.selectedService = true
+
+            previousService?.let {
+                it.selectedService = false
+                it.stopService()
+            }
+        } else {
+            // this one worked, but we already selected a service
+            // TODO use these later?
+            Log.d(TAG, "additional working service $test")
+            additionalWorkingConnections.add(test)
```

Comment:
should we stop the services for redundant successes or leave them running to possibly use later?

Reply:
We should stop all the services we're not using because: we're not using others in any way yet, it's potentially wasting resources just running in the background (some services maintain an active connection to the server), and I assume if and when we get to using the additional services, we'll want to re-test them. E.g. if it's an hour later or something, we can't just assume it still works, but it would be a good first one to try